README.MD
<p align="center"><img width="200" src="images/logo.png"></p>

## Introduction

Valet+ is a development environment for macOS. No Vagrant, no Docker, no `/etc/hosts` file. This project uses 
[laravel/valet](https://github.com/laravel/valet) as a dependency.


## Installation

> :warning: Valet+ requires macOS and [Homebrew](https://brew.sh/). Before installation, you should make sure that no 
> other programs such as Apache or Nginx are binding to your local machine's port 80.

> :warning: Valet+ requires [Composer](https://getcomposer.org/) to be installed. 

To get started, you first need to ensure that Homebrew is up to date using the update command:
```shell
brew update
```

Next, you should use Homebrew to install PHP 8.0:
```shell
brew install php@8.0
brew link php@8.0 --force
brew services restart php@8.0
```

Install Valet+ 3.x-dev with composer:
```shell
composer global require weprovide/valet-plus:3.x-dev
```

Finally, you may execute the Valet+ `install` command. This will configure and install Valet+, Nginx, DnsMasq and other
services. In addition, use `--help` to see all installation options. 
```shell
valet-plus install
```

The daemons Valet+ depends on will be configured to launch when your system starts. Once Valet+ is installed, try pinging
any `*.test` domain on your terminal using a command such as `ping foobar.test`. If Valet+ is installed correctly you 
should see this domain responding on `127.0.0.1`.



## Valet+ features

Here are a few key differences compared to the original Valet:

- Nginx config optimization
- PHP extensions (~~mcrypt~~, ~~intl~~, ~~opcache~~, yaml, apcu)
- MySQL (mysql@5.7, mysql@8.0, mysql@8.1, mariadb)
  - DB commands (list, create, drop, reset, (re)import, export, ~~open~~)
- Mailhog (on/off mode)
- Varnish (on/off mode)
- Redis (on/off mode)
- Rabbitmq (on/off mode)
- Xdebug (on/off mode)
- Memcache (on/off mode)
- Elasticsearch v6, v7, ~~v8~~ (on/off mode) using Docker 
- Opensearch (on/off mode)
- ~~Ioncube~~
- Rewrite/unrewrite public domain to local environment
- ~~DevTools~~
- Binaries (magerun, magerun2, drush, wp-cli)

### Changes vs Valet+ 2

- Use command `valet-plus` instead of `valet` (for now).
- Rename `.env.valet` to `.valet-env.php`.
- Use command `valet-plus elasticsearch|es use <version>` instead of `valet-plus use elasticsearch|es <version>`.
- Use `127.0.0.1` as Redis host instead of `/tmp/redis.sock`.
- Choose which binaries to install (default all) and self-update on `valet-plus install` command.
- Adds dependency on Docker for Elasticsearch, see https://docs.docker.com/desktop/install/mac-install/



## Valet+ Authors

- Mischa Braam ([@mischabraam](https://github.com/mischabraam))
- Tim Neutkens ([@timneutkens](https://github.com/timneutkens))
- Sam Granger ([@samgranger](https://github.com/samgranger))

## Valet+ Contributors

- Lou van der Laarse ([@Neodork](https://github.com/Neodork))
- Sander Pott (@Sander Valet+ slack channel)



## Get in touch

[![slack](https://p9.zdassets.com/hc/theme_assets/138842/200037786/logo.png)](https://join.slack.com/t/valet-plus/shared_invite/enQtNDE2MjU2NzgyNjQwLWFiYWNjOWFhOWQ2ZDcyOTEyZTA2MzAzOWYyYzYwMTYzODVlMGE3ZDg3ZWQ1M2JmN2M0OGY3OGUwMDI3NDM1NDU)

We have a slack workspace available [which you can join](https://join.slack.com/t/valet-plus/shared_invite/enQtNDE2MjU2NzgyNjQwLWFiYWNjOWFhOWQ2ZDcyOTEyZTA2MzAzOWYyYzYwMTYzODVlMGE3ZDg3ZWQ1M2JmN2M0OGY3OGUwMDI3NDM1NDU).

Yes, we have a Valet+ fan page https://www.weprovide.com/valet-plus


CHANGELOG.md
# Changelog
All notable changes to valet-plus will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.4.2]
### Fixed
- [#627] Add support for PHP 8.3
- [#478] Fix not working elasticsearch.test domain

## [2.4.1]
### Fixed
- [#621] Fixed ngrok script

## [2.4.0](https://github.com/weprovide/valet-plus/compare/2.3.1...2.4.0)
### Fixed
- [#612] Fixed Bedrock driver
- [#522] Use SHA256 for self signed certificates
- [#608] Update xdebug to support PHP 8.1

### Added
- Support for PHP 8.1
- Support for PHP 8.2

### Changed
- Replaced valet-php tap by shivammathur php tap
- Replaced pecl extensions by shivammathur php extensions

### Removed
- Support for PHP <7.1


## [2.3.1](https://github.com/weprovide/valet-plus/compare/2.3.0...2.3.1)
### Fixed
- [#596] Fix mysql 8 installation
- [#595] Fix opening log with dynamic Brew path

### Added
- [#594] Add tailing logs

## [2.3.0](https://github.com/weprovide/valet-plus/compare/2.2.1...2.3.0)
### Fixed
- [#587] Minimum version of symfony/process causes fatal error
- [#549] Fix xdebug veriosn 2.9.8 for PHP 7.x

### Changed
- Start highest installed major of Elasticsearch on valet start
- Use PHP 7.4 as default version

### Added
- [#572] Add support for usage on ARM processors (M1 mac's) 
- [#563] Add support for Xdebug 3
- Add libyaml as brew devtool
- [#567] Add support for Mysql 8

## [2.2.1](https://github.com/weprovide/valet-plus/compare/2.2.0...2.2.1)
- Fix missing dependencies of sebastian/version

## [2.2.0](https://github.com/weprovide/valet-plus/compare/2.1.0...2.2.0)
- [#517] Nginx config changes for fastcgi_buffers
- [#554] Fix for Brew 2.7.0: brew list without commands
- [#551] Replace hardcoded Elasticsearch versions by fetching from Brew
- [#546] Add support for PHP 8.0
- [#583] Add support for ioncube on PHP 7.4
- Add --build-from-source when installing PHP
- Use PHP 7.2 as default version

## [2.1.0](https://github.com/weprovide/valet-plus/compare/2.0.0...2.1.0)
### Removed
- [#524] Removed support for Elasticsearch 2.4
- [#524] Removed support for Elasticsearch 5.6
- [#524] Removed support for Elasticsearch 7.6
- [#516] Removed setting the query parameter for Drupal.
- [#535] Removed old PHP fixes within valet fix for old brew formulae now replaced by henkrehost/php

### Added
- [#524] Added support for Elasticsearch 7.8
- [#502] Added to PHP xdebug config xdebug.max_nesting_level to -1
- [#530] Added Composer.lock file
- [#487] Added brew update and uninstall brew php pipeline step
- [#470] Added macOS catalina image for testing on macos catalina

### Changed
- [#501] Changed PHP memory limit to 4 GB
- [#454] Changed the valet unlink command to improve feedback and help description.

### Fixed
- [#541] Fixed PECL YML package no longer supporting PHP 7.0 beyond version 2.0.4.
- [#539] Fixed the CI/CD not failing when switch to (php) step fails 
- [#537] Fixed valet fix forcefully trying to reinstall and relink the default PHP version. 
Even when not on the default version. Causing a faulty link process (overwriting).
- [#533] Fixed CI/CD not adhering to installation instructions
- [#528] Fixed syntax error due to symfony/polyfill-php80 being required by new version (4.4.9) of symfony/debug
- [#530] Fixed composer.lock being ignored causing composer updates to run during install
- [#464] Fixed logic to copy Elasticsearch NGINX stub on install
- [#455] Fixed bug where switching PHP version to a flawed installation would result in an unusable valet-plus installation.

## [2.0.0](https://github.com/weprovide/valet-plus/compare/1.0.29...2.0.0)
### Removed
- [#393] Major part of the readme, now available in the [WIKI](https://github.com/weprovide/valet-plus/wiki)
- [#437] Dependency on mysql-utilities because brew has deprecated the use of the formula.

### Fixed 
- [#401] Installation bug with apcu_bc due to double .so directives getting placed in php.ini.
- [#402] Installation bug with ioncube due to PHP 7.3 not being configured.
- [#403] Installation bug with memcached on MacOS Mojave due to zlib not being installed by default anymore.
- [#408] Installation bug with Elasticsearch due to PECL yaml not being installed.
- [#408] Bug while switching Elasticsearch version by always suffixing the version in the datapath config.
- [#425] Bug in custom PECL extensions throwing errors when no url is set for custom extension.
- [#426] Bug in valet fix command trying to install PHP 7.1 as default.

cli/Valet/Pecl.php
<?php

namespace Valet;

use Exception;
use DomainException;

/**
 * @deprecated
 */
class Pecl extends AbstractPecl
{

    // Extensions.
    const XDEBUG_EXTENSION = 'xdebug';
    const APCU_EXTENSION = 'apcu';
    const APCU_BC_EXTENSION = 'apcu_bc';
    const GEOIP_EXTENSION = 'geoip';
    const MEMCACHE_EXTENSION = 'memcached';
    const YAML_EXTENSION = 'yaml';

    // Extension aliases.
    const APCU_BC_ALIAS = 'apc';

    /**
     * Supported pecl extensions for the PECL manager. There are 2 types of extensions supported: zend_extensions and
     * extensions. Not all extension are available through PECL. This class handles packages that are handled by PECL.
     *
     * PECL example:
     *
     * The PECL extensions sometimes differ in version per PHP version. For example Xdebug can only be installed on
     * PHP 5.6 using the 2.2.7 version. In such a case one would define the version within the array as key. The value
     * would be 2.2.7 which is the version to be installed. See example below.
     *
     * The 'default' key allows one to disable the module by default. This is useful for modules that enable through
     * on/off commands, like Xdebug.
     *
     * @formatter:off
     *
     * 'extension_key_name' => [
     *    '5.6' => '2.2.7'
     *    'default' => false
     * ]
     *
     * @formatter:on
     */
    const EXTENSIONS = [
        self::XDEBUG_EXTENSION => [
            '8.3' => '3.3.0',
            '8.2' => '3.1.5',
            '8.1' => '3.1.5',
            '8.0' => '3.1.5',
            '7.4' => '3.1.5',
            '7.3' => '3.1.5',
            '7.2' => '2.9.8',
            '7.1' => '2.9.8',
            '7.0' => '2.9.8',
            '5.6' => '2.2.7',
            'default' => false,
            'extension_type' => self::ZEND_EXTENSION_TYPE
        ],
        self::APCU_EXTENSION => [
            '7.3' => '5.1.17',
            '7.2' => '5.1.17',
            '7.1' => '5.1.17',
            '7.0' => '5.1.17',
            '5.6' => '4.0.11',
            'extension_type' => self::NORMAL_EXTENSION_TYPE
        ],
        self::GEOIP_EXTENSION => [
            '8.3' => false, //todo; will probably be 1.1.2
            '8.2' => false, //todo; will probably be 1.1.2
            '8.1' => false, //todo; will probably be 1.1.2
            '8.0' => false, //todo; will probably be 1.1.2
            '7.4' => '1.1.1',
            '7.3' => '1.1.1',
            '7.2' => '1.1.1',
            '7.1' => '1.1.1',
            '7.0' => '1.1.1',
            'extension_type' => self::NORMAL_EXTENSION_TYPE,
            'brew_dependency' => 'geoip'
        ],
        self::MEMCACHE_EXTENSION => [
            '7.3' => '3.1.3',
            '7.2' => '3.1.3',
            '7.1' => '3.1.3',
            '7.0' => '3.1.3',
            'default' => false,
            'extension_type' => self::NORMAL_EXTENSION_TYPE
        ],
        self::YAML_EXTENSION => [
            '5.6' => '1.3.1',
            '7.0' => '2.0.4',
            'extension_type' => self::NORMAL_EXTENSION_TYPE,
            'brew_dependency' => 'libyaml'
        ]
    ];

    /**
     * @var PeclCustom
     */
    public $peclCustom;
    /**
     * @var Architecture
     */
    private $architecture;

    /**
     * @inheritdoc
     */
    public function __construct(Architecture $architecture, CommandLine $cli, Filesystem $files, PeclCustom $peclCustom)
    {
        parent::__construct($cli, $files);
        $this->peclCustom = $peclCustom;
        $this->architecture = $architecture;
    }

    /**
     * @inheritdoc
     */
    public function installExtensions($onlyDefaults = true)
    {
        info("[PECL] Installing extensions");
        foreach (self::EXTENSIONS as $extension => $versions) {
            if ($onlyDefaults && $this->isDefaultExtension($extension) === false) {
                continue;
            }

            if ($this->getVersion($extension) !== false) {
                $this->installExtension($extension);
                $this->enableExtension($extension);
            }
        }
    }

    /**
     * Install a single extension if not already installed, default is true and has a version for this PHP version.
     * If version equals false the extension will not be installed for this PHP version.
     *
     * @param $extension
     *    The extension key name.
     * @return bool
     */
    public function installExtension($extension)
    {
        if ($this->isInstalled($extension)) {
            output("\t$extension is already installed, skipping...");
            return false;
        }

        $this->install($extension, $this->getVersion($extension));
        return true;
    }

    /**
     * Install a single extension.
     *
     * @param $extension
     *    The extension key name.
     * @param null $version
     */
    protected function install($extension, $version = null)
    {
        if ($version === null) {
            if ($this->hasBrewDependency($extension)) {
                $result = $this->cli->runAsUser("echo $(brew --prefix " . $this->getBrewDependency($extension) . ") | pecl install $extension");
            } else {
                $result = $this->cli->runAsUser("pecl install $extension");
            }
        } else {
            $result = $this->cli->runAsUser("pecl install $extension-$version");
        }

        $alias = $this->getExtensionAlias($extension);
        if (!preg_match("/Installing '(.*$alias.so)'/", $result)) {
            throw new DomainException("Could not find installation path for: $extension\n\n$result");
        }

        if (strpos($result, "Error:")) {
            throw new DomainException("Installation path found, but installation failed:\n\n$result");
        }

        $phpIniPath = $this->getPhpIniPath();
        $phpIniFile = $this->files->get($phpIniPath);
        if (!preg_match('/(zend_extension|extension)\="(.*' . $alias . '.so)"/', $phpIniFile, $iniMatches)) {
            throw new DomainException("Could not find ini definition for: $extension in $phpIniPath");
        }

        output("\t$extension successfully installed");
    }

    /**
     * Enable an single extension if not already enabled, default is true and has a version does not equal false.
     *
     * @param $extension
     *    The extension key name.
     * @return bool
     */
    public function enableExtension($extension)
    {
        if ($this->isEnabled($extension) && $this->isEnabledCorrectly($extension)) {
            output("\t$extension is already enabled, skipping...");
            return false;
        }

        $this->enable($extension);
        return true;
    }

    /**
     * Enable an single extension.
     *
     * @param $extension
     *    The extension key name.
     */
    public function enable($extension)
    {
        $phpIniPath = $this->getPhpIniPath();
        $phpIniFile = $this->files->get($phpIniPath);
        $phpIniFile = $this->replaceIniDefinition($extension, $phpIniFile);
        $phpIniFile = $this->alternativeInstall($extension, $phpIniFile);
        $this->peclCustom->saveIniFile($phpIniPath, $phpIniFile);

        output("\t$extension successfully enabled");
    }

    /**
     * @inheritdoc
     */
    public function uninstallExtensions()
    {
        info("[PECL] Removing extensions");
        foreach (self::EXTENSIONS as $extension => $versions) {
            if ($this->getVersion($extension) !== false) {
                $this->disableExtension($extension);
            }
        }
    }

    /**
     * Uninstall and disable an extension if installed.
     *
     * @param $extension
     *    The extension key name.
     * @return bool
     *    Whether or not an uninstall happened.
     */
    public function disableExtension($extension)
    {
        $version = $this->getVersion($extension);
        if ($this->isEnabled($extension)) {
            $this->disable($extension);
        }
        if ($this->isInstalled($extension)) {
            $this->uninstall($extension, $version);
            return true;
        }
        return false;
    }

    public function disable($extension)
    {
        $this->removeIniDefinition($extension);
        $this->alternativeDisable($extension);
    }

    /**
     * Replace and remove all directives of the .so file for the given extension within the php.ini file.
     *
     * @param $extension
     *    The extension key name.
     */
    private function removeIniDefinition($extension)
    {
        $phpIniPath = $this->getPhpIniPath();
        $alias = $this->getExtensionAlias($extension);
        $phpIniFile = $this->files->get($phpIniPath);
        $phpIniFile = preg_replace('/;?(zend_extension|extension)\=".*' . $alias . '.so"/', '', $phpIniFile);
        $this->peclCustom->saveIniFile($phpIniPath, $phpIniFile);
        output("\t$extension successfully disabled");
    }

    /**
     * Because some extensions install others, like apcu_bc, the default disable method will not be sufficient.
     * In that case use this method to for example add additional dependencies for the specific disable. For example
     * apcu_bc installs apcu and apc both will need to be removed from the php.ini file. The default would only remove
     * apcu.so so we define apc.so here as alternative.
     *
     * @param $extension
     * @return string
     */
    private function alternativeDisable($extension)
    {
        switch ($extension) {
            default:
                break;
        }
    }

    /**
     * Because some extensions install others, like apcu_bc, the default uninstall method will not be sufficient.
     * In that case use this method to for example add additional dependencies for the specific uninstall. For example
     * apcu_bc installs apcu and apc both will need to be uninstalled from pecl. The default would only uninstall
     * apcu.so so we define apc.so here as alternative.
     *
     * @param $extension
     * @return string
     */
    private function alternativeUninstall($extension)
    {
        switch ($extension) {
            default:
                break;
        }
    }

    /**
     * Uninstall a single extension.
     *
     * @param $extension
     *    The extension key name.
     * @param null $version
     */
    private function uninstall($extension, $version = null)
    {
        if ($version === null || $version === false) {
            $this->cli->passthru("pecl uninstall $extension");
        } else {
            $this->cli->passthru("pecl uninstall $extension-$version");
        }

        $this->alternativeUninstall($extension);
    }

    /**
     * Update the default PECL channel.
     */
    public function updatePeclChannel()
    {
        info('[PECL] Updating PECL channel: pecl.php.net');
        $this->cli->runAsUser('pecl channel-update pecl.php.net');
    }

    /**
     * Fix common problems related to the PECL/PEAR installation.
     */
    public function fix()
    {
        info('[PECL] Checking pear config...');
        // Check if pear config is set correctly as per:
        // https://github.com/kabel/homebrew-core/blob/2564749d8f73e43cbb8cfc449bca4f564ac0e9e1/Formula/php%405.6.rb
        // Brew installation standard.
        foreach (PhpFpm::SUPPORTED_PHP_FORMULAE as $phpVersion => $brewname) {
            output("Checking php $phpVersion...");

            $pearConfigPath = $this->architecture->getBrewPath() . PhpFpm::LOCAL_PHP_FOLDER . "$phpVersion/pear.conf";

            if (!$this->files->exists($pearConfigPath)) {
                warning("    Skipping $phpVersion, Pear config path could not be found at: $pearConfigPath");
                continue;
            }

            $pearConfig = $this->files->get($pearConfigPath);
            $pearConfigSplit = explode("\n", $pearConfig);

            $pearConfigVersion = '';
            $pearConfig = false;
            foreach ($pearConfigSplit as $splitValue) {
                if (strpos($splitValue, 'php_ini')) {
                    $pearConfig = unserialize($splitValue);
                } else if (strpos($splitValue, 'Config')) {
                    $pearConfigVersion = $splitValue;
                } else {
                    continue;
                }
            }

            if ($pearConfig === false) {
                warning("Could not determine pear configuration for PhP version: $phpVersion, skipping...");
                continue;
            }

            $pearName = $this->replacePhpWithPear($brewname);

            $phpIniPath = str_replace('pear.conf', 'php.ini', $pearConfigPath);
            $phpDirPath = $this->architecture->getBrewPath() . "/share/$pearName";
            $pearDocDirPath = $this->architecture->getBrewPath() . "/share/$pearName/doc";
            $phpExtensionDirPath = $this->architecture->getBrewPath() . '/lib/php/pecl/'.basename($pearConfig['ext_dir']);
            $phpBinPath = $this->architecture->getBrewPath() . "/opt/$brewname/bin";
            $pearDataDirPath = $this->architecture->getBrewPath() . "/share/$pearName/data";
            $pearCfgDirPath = $this->architecture->getBrewPath() . "/share/$pearName/cfg";
            $pearWwwDirPath = $this->architecture->getBrewPath() . "/share/$pearName/htdocs";
            $pearManDirPath = $this->architecture->getBrewPath() . '/share/man';
            $pearTestDirPath = $this->architecture->getBrewPath() . "/share/$pearName/test";
            $phpBinDirPath = $this->architecture->getBrewPath() . "/opt/$brewname/bin/php";

            // Check php_ini value of par config.
            if (empty($pearConfig['php_ini']) || $pearConfig['php_ini'] !== $phpIniPath) {
                output("    Setting pear config php_ini directive to: $phpIniPath");
                $pearConfig['php_ini'] = $phpIniPath;
            }
            // Check php_dir value of par config.
            if (empty($pearConfig['php_dir']) || $pearConfig['php_dir'] !== $phpDirPath) {
                output("    Setting pear config php_dir directive to: $phpDirPath");
                $pearConfig['php_dir'] = $phpDirPath;
            }
            // Check doc_dir value of par config.
            if (empty($pearConfig['doc_dir']) || $pearConfig['doc_dir'] !== $pearDocDirPath) {
                output("    Setting pear config doc_dir directive to: $pearDocDirPath");
                $pearConfig['doc_dir'] = $pearDocDirPath;
            }
            // Check ext_dir value of par config.
            if (empty($pearConfig['ext_dir']) || $pearConfig['ext_dir'] !== $phpExtensionDirPath) {
                output("    Setting pear config ext_dir directive to: $phpExtensionDirPath");
                $pearConfig['ext_dir'] = $phpExtensionDirPath;
            }
            // Check php_bin value of par config.
            if (empty($pearConfig['bin_dir']) || $pearConfig['bin_dir'] !== $phpBinPath) {
                output("    Setting pear config bin_dir directive to: $phpBinPath");
                $pearConfig['bin_dir'] = $phpBinPath;
            }
            // Check data_dir value of par config.
            if (empty($pearConfig['data_dir']) || $pearConfig['data_dir'] !== $pearDataDirPath) {
                output("    Setting pear config data_dir directive to: $pearDataDirPath");
                $pearConfig['data_dir'] = $pearDataDirPath;
            }
            // Check cfg_dir value of par config.
            if (empty($pearConfig['cfg_dir']) || $pearConfig['cfg_dir'] !== $pearCfgDirPath) {
                output("    Setting pear config cfg_dir directive to: $pearCfgDirPath");
                $pearConfig['cfg_dir'] = $pearCfgDirPath;
            }
            // Check www_dir value of par config.
            if (empty($pearConfig['www_dir']) || $pearConfig['www_dir'] !== $pearWwwDirPath) {
                output("    Setting pear config www_dir directive to: $pearWwwDirPath");
                $pearConfig['www_dir'] = $pearWwwDirPath;
            }
            // Check man_dir value of par config.
            if (empty($pearConfig['man_dir']) || $pearConfig['man_dir'] !== $pearManDirPath) {
                output("    Setting pear config man_dir directive to: $pearManDirPath");
                $pearConfig['man_dir'] = $pearManDirPath;
            }
            // Check test_dir value of par config.
            if (empty($pearConfig['test_dir']) || $pearConfig['test_dir'] !== $pearTestDirPath) {
                output("    Setting pear config test_dir directive to: $pearTestDirPath");
                $pearConfig['test_dir'] = $pearTestDirPath;
            }
            // Check php_bin value of par config.
            if (empty($pearConfig['php_bin']) || $pearConfig['php_bin'] !== $phpBinDirPath) {
                output("    Setting pear config php_bin directive to: $phpBinDirPath");
                $pearConfig['php_bin'] = $phpBinDirPath;
            }

            // Rebuild the config.
            $pearConfig = serialize($pearConfig);
            if (!empty($pearConfigVersion)) {
                $pearConfig = $pearConfigVersion."\n".$pearConfig;
            }

            // Put config back into the pear.conf file.
            $this->files->put($pearConfigPath, $pearConfig);
        }
    }

    /**
     * @inheritdoc
     */
    public function isInstalled($extension)
    {
        return strpos($this->cli->runAsUser('pecl list | grep ' . $extension), $extension) !== false;
    }

    /**
     * Because some extensions install others, like apcu_bc, the default install method will not be sufficient.
     * In that case use this method to for example add additional dependencies for the specific install. For example
     * apcu_bc installs apcu and apc both will need to be linked within the php.ini file. The default would only link
     * apcu.so so we define apc.so here as alternative.
     *
     * @param $extension
     *    The extension key name.
     * @param $phpIniFile
     * @return string
     */
    private function alternativeInstall($extension, $phpIniFile)
    {
        switch ($extension) {
            default:
                return $phpIniFile;
        }
    }

    /**
     * Replace the .so directive within the php.ini file. Initial .so directive will be set by
     * 'pecl install {extension}'. Replacing will prevent duplicate entries.
     *
     * @param $extension
     *    The extension key name.
     * @param $phpIniFile
     *
     * @return string
     */
    private function replaceIniDefinition($extension, $phpIniFile)
    {
        $alias = $this->getExtensionAlias($extension);
        $type = $this->getExtensionType($extension);
        $phpIniFile = preg_replace("/(zend_extension|extension)\=\"(.*$alias.so)\"/", '', $phpIniFile);

        return "$type=\"$alias.so\"\n" . $phpIniFile;
    }

    /**
     * Whether or not this extension should be enabled by default. Is set by setting the
     * 'default' key to true within the extensions configuration.
     *
     * @param $extension
     *    The extension key name.
     * @return bool
     */
    private function isDefaultExtension($extension)
    {
        if (array_key_exists('default', self::EXTENSIONS[$extension])) {
            return false;
        } elseif (array_key_exists('default', self::EXTENSIONS[$extension]) === false) {
            return true;
        } else {
            return false;
        }
    }

    /**
     * @inheritdoc
     */
    protected function getExtensionAlias($extension)
    {
        switch ($extension) {
            default:
                return $extension;
        }
    }

    /**
     * Get the version of the extension supported by the current PHP version.
     *
     * @param $extension
     *    The extension key name.
     * @return null
     */
    private function getVersion($extension)
    {
        $phpVersion = $this->getPhpVersion();
        if (array_key_exists($phpVersion, self::EXTENSIONS[$extension])) {
            return self::EXTENSIONS[$extension][$phpVersion];
        }
        return null;
    }

    /**
     * Check if the extension has any brew dependency
     *
     * @param mixed $extension
     * @return bool
     */
    private function hasBrewDependency($extension)
    {
        return array_key_exists("brew_dependency", self::EXTENSIONS[$extension]);
    }


    /**
     * Get the brew dependency for the pecl command
     *
     * @param mixed $extension
     * @return mixed
     */
    private function getBrewDependency($extension)
    {
        return self::EXTENSIONS[$extension]["brew_dependency"];
    }

    /**
     * Pecl sometimes adds a directive twice. This causes the PHP installation to see the extension as
     * installed. However the seconds directive is seen as faulty which causes the PHP installation to
     * fail. This van be solved by checking if the directive exists twice. If it does valet-plus will
     * deem the module as "disabled" and replace the existing directives with a single one.
     *
     * @param $extension
     * @return bool
     */
    private function isEnabledCorrectly($extension)
    {
        $phpIniPath = $this->getPhpIniPath();
        $phpIniFile = $this->files->get($phpIniPath);
        $type = $this->getExtensionType($extension);
        $alias = $this->getExtensionAlias($extension);
        preg_match_all("/$type=\"$alias.so\"/m", $phpIniFile, $matches);
        $alternativeEnabledCorrectly = $this->isAlternativeEnabledCorrectly($extension);
        $isEnabledCorrectly = $alternativeEnabledCorrectly && (is_array($matches) && count($matches) === 1?count($matches[0]) === 1:false);
        return $isEnabledCorrectly;
    }

    /**
     * Alternative case for checking directives of modules that depend on other modules.
     *
     * @param $extension
     * @return bool
     */
    private function isAlternativeEnabledCorrectly($extension)
    {
        switch ($extension) {
            default:
                return true;
        }
    }

    private function replacePhpWithPear($brewname)
    {
        return str_replace('php', 'pear', $brewname);
    }
}
